utils/common.cpp

```diff
@@ -174,14 +174,32 @@ static ssize_t write_n(int fd, const void *buf, size_t n) {
   return nwritten;
 }
 
-void ReadFile(int fd, void *buf, u32 len, bool original_behaviour) {
-  if (read_n(fd, buf, len, original_behaviour) != len)
-    throw FileError(StrPrintf("Failed to read fd=%d", fd));
+// Read exact `len` bytes
+ssize_t ReadFile(int fd, void *buf, u32 len, bool original_behaviour) {
+  try {
+    ssize_t nbytes = read_n(fd, buf, len, original_behaviour);
+    if (nbytes != len) {
+      if (original_behaviour)
+        return -1;
+      else 
+        throw FileError(StrPrintf("Failed to read exact len bytes: fd=%d, len=%d, nbytes=%d", fd, len, nbytes));
+    }
+    return nbytes;
+  } catch (std::system_error& e) {
+    throw FileError(StrPrintf("Failed to read fd=%d: errno_to_system_error(errno=%s)", fd, e.what()));
+  }
```

Comment:
We don't want to use try-catch statement because it's makes the program slow, especially in this kind of utility that will be called many times.
I guess you can fix it like this:
```suggestion
  ssize_t nbytes = read_n(fd, buf, len, true);
  if (!original_behaviour && nbytes != len) {
    throw FileError(StrPrintf("Failed to read exact len bytes: fd=%d, len=%d, nbytes=%d", fd, len, nbytes));
  }
  return nbytes;
```

Reply:
The original behaviour of `ReadFile` as far as I understand is:
- If `original_behaviour` is not set, `original_behaviour` is true
- If `original_behaviour` is true:
  - If read works, return `nbytes`
  - If read fails, throw `FileError`
- If `original_behaviour` is false:
  - If read works, return `nbytes`
  - If read fails, throw `system_error` (in `read_n`)

My new suggestion works as
- If `original_behaviour` is not set, `original_behaviour` is **false**
- `ReadFile` calls `read_n` always with `original_behaviour` set to true
  - If read works, return `nbytes`
  - If read fails:
    - If `original_behaviour` is true, return -1
    - If `original_behaviour` is false, throw `FileError`

Reply:
Thak you for suggestion. 
I applied your suggestion with minor modificatgion.
Patched code looks good. 
Resolved.